lib/route/nh_encap_seg6_local.c

```diff
@@ -0,0 +1,1161 @@
+	int i;
+
+	slwt_a = a;
+	slwt_b = b;
+
+	if (slwt_a->action != slwt_b->action)
+		return 1;
+
+	attrs_a = slwt_a->desc->attrs | slwt_a->parsed_optattrs;
+	attrs_b = slwt_b->desc->attrs | slwt_b->parsed_optattrs;
+
+	if (attrs_a != attrs_b)
+		return 1;
+
+	for (i = SEG6_LOCAL_SRH; i < SEG6_LOCAL_MAX + 1; i++) {
+		if (attrs_a & SEG6_F_ATTR(i)) {
+			param = &seg6_action_params[i];
+			if (param->cmp(slwt_a, slwt_b))
+				return 1;
+		}
+	}
+
+	return 0;
+}
+
+static const char *seg6_action_names[SEG6_LOCAL_ACTION_MAX + 1] = {
+	[SEG6_LOCAL_ACTION_END]			= "End",
+	[SEG6_LOCAL_ACTION_END_X]		= "End.X",
+	[SEG6_LOCAL_ACTION_END_T]		= "End.T",
+	[SEG6_LOCAL_ACTION_END_DX2]		= "End.DX2",
+	[SEG6_LOCAL_ACTION_END_DX6]		= "End.DX6",
+	[SEG6_LOCAL_ACTION_END_DX4]		= "End.DX4",
+	[SEG6_LOCAL_ACTION_END_DT6]		= "End.DT6",
+	[SEG6_LOCAL_ACTION_END_DT4]		= "End.DT4",
+	[SEG6_LOCAL_ACTION_END_B6]		= "End.B6",
+	[SEG6_LOCAL_ACTION_END_B6_ENCAP]	= "End.B6.Encaps",
+	[SEG6_LOCAL_ACTION_END_BM]		= "End.BM",
+	[SEG6_LOCAL_ACTION_END_S]		= "End.S",
+	[SEG6_LOCAL_ACTION_END_AS]		= "End.AS",
+	[SEG6_LOCAL_ACTION_END_AM]		= "End.AM",
+	[SEG6_LOCAL_ACTION_END_BPF]		= "End.BPF",
+	[SEG6_LOCAL_ACTION_END_DT46]		= "End.DT46",
+};
+
+static const char *format_action_type(int action)
+{
+	if (action < 0 || action > SEG6_LOCAL_ACTION_MAX)
+		return "<invalid>";
+
+	return seg6_action_names[action] ?: "<unknown>";
+}
+
+static void seg6_local_encap_dump(void *priv, struct nl_dump_params *dp)
+{
+	struct seg6_local_lwt *slwt;
+	struct seg6_action_param *param;
+	unsigned long attrs;
+	int i;
+
+	slwt = priv;
+	nl_dump(dp, "action %s ", format_action_type(slwt->action));
+
+	attrs = slwt->desc->attrs | slwt->parsed_optattrs;
+
+	for (i = SEG6_LOCAL_SRH; i < SEG6_LOCAL_MAX + 1; i++) {
+		if (attrs & SEG6_F_ATTR(i)) {
+			param = &seg6_action_params[i];
+			if (param->dump)
+				param->dump(slwt, dp);
+		}
+	}
+}
+
+struct nh_encap_ops seg6_local_encap_ops = {
+	.encap_type	= LWTUNNEL_ENCAP_SEG6_LOCAL,
+	.build_msg	= seg6_local_encap_build_msg,
+	.parse_msg	= seg6_local_encap_parse_msg,
+	.compare	= seg6_local_encap_compare,
+	.dump		= seg6_local_encap_dump,
+	.destructor	= seg6_local_encap_destructor,
+};
+
+static struct seg6_local_lwt *
+get_seg6_local_slwt(struct rtnl_nexthop *nh)
+{
+	if (!nh->rtnh_encap || nh->rtnh_encap->ops->encap_type != LWTUNNEL_ENCAP_SEG6_LOCAL)
+		return NULL;
+
+	return (struct seg6_local_lwt *)nh->rtnh_encap->priv;
+}
+
+int rtnl_route_nh_get_encap_seg6_local_action(struct rtnl_nexthop *nh)
+{
+	struct seg6_local_lwt *slwt;
+
+	slwt = get_seg6_local_slwt(nh);
+	if (slwt == NULL)
+		return 0;
+
+	return slwt->action;
+}
+
+int rtnl_route_nh_has_encap_seg6_local_attr(struct rtnl_nexthop * nh, int attr)
+{
+	unsigned long attrs;
+	struct seg6_local_lwt *slwt;
+
+	slwt = get_seg6_local_slwt(nh);
+	if (slwt == NULL)
+		return 0;
+
+	attrs = slwt->desc->attrs | slwt->parsed_optattrs;
+
+	return attrs & SEG6_F_ATTR(attr);
```

Comment:
`SEG6_F_ATTR(attr)` does a bitshift based on `attr`. attr is user provided input, let's check that the `attr` argument is suitably small (and non-negative), to not cause undefined behavior.

```C
if (attr < 0 or attr >= sizeof(attr)*8) return -NLE_INVAL;
```

But more importantly, please return a negative error code
- `attr` is invalid -> `-NLE_INVAL`
- `if (!slwt)` -> `-NLE_NOATTR`

otherwise, return either 0 or 1, via `return !!(attrs & SEG6_F_ATTR(attr));`

A doctext explaining the purpuse/use of the function and behavior (return values) would also be necessary.